### ISSUE TITLE
support the newer zstd compression

### ISSUE THREAD

**Issue by bobombolo** [LeMagnesium/python-minetest#19]
i heard on the minetest forums that this library doesn't support the latest world sqlite format -- something to do with zstd compression?

**Comment by Lymkwi** [comment]
It's not entirely surprising that my library doesn't support zstd compression. I wrote it more than half a decade ago (it's bad Python, the kind you write when you haven't been coding for long), and that's about when I last opened the game as well.

Are people still using this? It could be worth re-opening good ol' MT for the lols and seeing if I can update my stuff easily to support newer versions.

**Comment by bobombolo** [comment]
i would play with it and i know someone on the MT forum is trying to make a Minecraft to Minetest map converter in python...

**Comment by Lymkwi** [comment]
Update: I'm working on a fix for map versions >= 29, but currently python's `zstd` is unable to handle the mapblock data in minetest's files.. I'll work a bit more on it. Shouldn't be too hard.

**Comment by teverett** [comment]
@Lymkwi did you ever solve the zstd issue?  I believe I am struggling with the same issue

**Comment by Lymkwi** [comment]
> @Lymkwi did you ever solve the zstd issue?  I believe I am struggling with the same issue

I think I managed to reproduce it but never figured anything useful out of the errors (or rather lack thereof) of Zstd. I had to dig in engine code to find the compression function and see if it did anything weird.. which it doesn't. Somehow it seems to be an incompatibility between python's Zstd and the library used by MT when both have default arguments.

Might try and give this a shot again, but I wouldn't expect a solution anytime soon

**Comment by teverett** [comment]
I did finally get it working.  The problem seemed to be how I was using Redis.  Once I had the right data, the decompression worked nicely.

**Comment by Lymkwi** [comment]
> The problem seemed to be how I was using Redis.

In what sense exactly?

**Comment by wgaylord** [comment]
> > @Lymkwi did you ever solve the zstd issue?  I believe I am struggling with the same issue
> 
> I think I managed to reproduce it but never figured anything useful out of the errors (or rather lack thereof) of Zstd. I had to dig in engine code to find the compression function and see if it did anything weird.. which it doesn't. Somehow it seems to be an incompatibility between python's Zstd and the library used by MT when both have default arguments.
> 
> Might try and give this a shot again, but I wouldn't expect a solution anytime soon


So I ran into this issue as well, seems they use the stream api of zstd so you need to use a Zstd library that supports that.

**Comment by Lymkwi** [comment]
I'm at a point where i think i no longer have any database for the game world anywhere but it's worth giving it a spin.

So instead of `{de,}compress`, my understanding is that i should use `stream_{reader,writer}` (from the correct class for compression/decompression), if i understand correctly and this is right

**Comment by wgaylord** [comment]
Yes, although it seems like the data is just the byte version number then the rest is zstd compressed.
